File: blue/src/colors_compute/lifecycle.py
"""Strict schema-two lifecycle journal intentions; no I/O or dispatch authority."""
from ._copy import deepcopy
import re

from .contract import _safe, state_keys
from .coordination import _identity, _shape, _nonblank, _integer, _role, _topology, _observation, MAX_INTEGER


def _fingerprint(value):
    return isinstance(value, str) and bool(re.fullmatch(r'SHA256:[A-Za-z0-9+/]{43}', value))


def _running(record):
    return record['phase'] in ('running', 'destroying')


def _idle_shared():
    return {'phase': 'declared', 'operation': None, 'operation_id': None}


def _record(record):
    phase = record['phase']
    if phase not in ('declared', 'running', 'ready', 'failed', 'destroying', 'destroyed'):
        return False
    if phase == 'declared':
        return record['operation'] is None and record['operation_id'] is None
    if not _safe(record['operation_id']) or record['operation'] not in ('create', 'destroy'):
        return False
    if phase in ('running', 'ready'):
        return record['operation'] == 'create'
    if phase in ('destroying', 'destroyed'):
        return record['operation'] == 'destroy'
    return True


def lifecycle_document_valid(value):
    if not _shape(value, ('schema_version', 'identity', 'revision', 'write_id', 'lock', 'generation', 'status', 'topology_declared', 'key', 'shared', 'nodes')):
        return False
    if not _integer(value['schema_version'], 2, 2) or not _identity(value['identity']) or not _integer(value['revision'], 1, MAX_INTEGER) or not _integer(value['generation'], 1, MAX_INTEGER) or not _safe(value['write_id']) or value['status'] not in ('active', 'deleting', 'retired') or type(value['topology_declared']) is not bool:
        return False
    lock = value['lock']
    if not _shape(lock, ('state', 'run_id')) or not ((lock['state'] == 'held' and _safe(lock['run_id'])) or (lock['state'] == 'idle' and lock['run_id'] is None)):
        return False
    key = value['key']
    if not _shape(key, ('mode', 'phase', 'fingerprint')) or key['phase'] not in ('absent', 'intent', 'prepared', 'cleanup', 'removed'):
        return False
    if key['phase'] == 'absent':
        if key['mode'] is not None or key['fingerprint'] is not None:
            return False
    else:
        if key['mode'] not in ('managed', 'external'):
            return False
        if key['mode'] == 'external' or key['phase'] == 'intent':
            if key['fingerprint'] is not None:
                return False
        elif not _fingerprint(key['fingerprint']):
            return False
    if not _shape(value['shared'], ('phase', 'operation', 'operation_id')) or not _record(value['shared']) or not isinstance(value['nodes'], dict) or len(value['nodes']) > 1000:
        return False
    operations = set()
    if value['shared']['operation_id'] is not None:
        operations.add(value['shared']['operation_id'])
    if _running(value['shared']) and lock['state'] != 'held':
        return False
    desired_roles = {}
    for node_id, node in value['nodes'].items():
        if not _safe(node_id) or not _shape(node, ('state_key', 'role', 'index', 'desired', 'phase', 'operation', 'operation_id')) or not _role(node['role']) or not _integer(node['index'], 0, 999) or type(node['desired']) is not bool or not _record(node):
            return False
        expected_id = str(int(node['index'])) if node['role'] is None else f"{node['role']}-{int(node['index'])}"
        if node_id != expected_id or node['state_key'] != f"{value['identity']['profile']}/compute/nodes/{node_id}.tfstate":
            return False
        if node['operation_id'] is not None:
            if node['operation_id'] in operations:
                return False
            operations.add(node['operation_id'])
        if _running(node) and lock['state'] != 'held':
            return False
        if node['desired']:
            desired_roles.setdefault(node['role'], []).append(int(node['index']))
    if None in desired_roles and len(desired_roles) != 1:
        return False
    destroyed = value['shared']['phase'] == 'destroyed' and all(node['phase'] == 'destroyed' for node in value['nodes'].values())
    undesired = all(not node['desired'] for node in value['nodes'].values())
    if value['status'] == 'retired' and not (key['phase'] == 'removed' and destroyed and undesired):
        return False
    if value['status'] == 'deleting' and not undesired:
        return False
    if key['phase'] in ('cleanup', 'removed') and not (value['status'] in ('deleting', 'retired') and destroyed):
        return False
    return all(sorted(indices) == list(range(len(indices))) for indices in desired_roles.values())


EXTRAS = {'acquire': (), 'release': (), 'begin-delete': (), 'retire': (), 'recreate': (),
          'declare': ('topology',), 'key-intent': ('mode',), 'key-prepared': ('fingerprint',),
          'key-cleanup': (), 'key-removed': (), 'shared-start': ('operation_id',),
          'shared-destroy': ('operation_id',), 'shared-complete': ('operation_id',),
          'shared-fail': ('operation_id',), 'shared-retry': ('evidence',),
          'start': ('node_id', 'operation_id'), 'destroy': ('node_id', 'operation_id'),
          'complete': ('node_id', 'operation_id'), 'fail': ('node_id', 'operation_id'),
          'retry': ('node_id', 'evidence')}


def _event(value):
    if not isinstance(value, dict) or not isinstance(value.get('type'), str) or not value['type'].startswith('lifecycle/'):
        return False
    name = value['type'][10:]
    if name not in EXTRAS or not _shape(value, ('type', 'run_id', 'write_id', 'target_etag', *EXTRAS[name])) or not _safe(value['run_id']) or not _safe(value['write_id']) or not (value['target_etag'] is None or _nonblank(value['target_etag'])):
        return False
    if 'node_id' in value and not _safe(value['node_id']):
        return False
    if 'operation_id' in value and not _safe(value['operation_id']):
        return False
    if name == 'declare' and _topology(value['topology']) is None:
        return False
    if name == 'key-intent' and value['mode'] not in ('managed', 'external'):
        return False
    if name == 'key-prepared' and value['fingerprint'] is not None and not _fingerprint(value['fingerprint']):
        return False
    return 'evidence' not in value or value['evidence'] == 'readable-state' or (name in ('shared-retry', 'retry') and value['evidence'] == 'verified-provider-absence')


def lifecycle(observation, identity, event):
    def fail(message):
        raise ValueError(message)
    def require(condition):
        if not condition:
            fail('lifecycle transition refused')
    if not _identity(identity):
        fail('invalid lifecycle identity')
    if not _event(event):
        fail('invalid lifecycle event')
    if not _observation(observation):
        fail('invalid lifecycle observation')
    if observation['status'] == 'error':
        fail('lifecycle read failed')
    present, doc, name = observation['status'] == 'present', observation.get('document'), event['type'][10:]
    if present and not lifecycle_document_valid(doc):
        fail('invalid lifecycle document')
    if present and doc['identity'] != identity:
        fail('lifecycle identity mismatch')
    if event['target_etag'] != (observation['etag'] if present else None):
        fail('lifecycle stale observation')
    if present and event['write_id'] == doc['write_id']:
        fail('lifecycle write_id reused')
    if present and doc['revision'] == MAX_INTEGER:
        fail('lifecycle revision exhausted')
    if not present:
        require(name == 'acquire')
        return {'condition': {'if_none_match': '*'}, 'document': {
            'schema_version': 2, 'identity': deepcopy(identity), 'revision': 1, 'write_id': event['write_id'],
            'lock': {'state': 'held', 'run_id': event['run_id']}, 'generation': 1, 'status': 'active',
            'topology_declared': False, 'key': {'mode': None, 'phase': 'absent', 'fingerprint': None},
            'shared': _idle_shared(), 'nodes': {}}}
    result = deepcopy(doc)
    def records():
        return [result['shared'], *result['nodes'].values()]
    def active():
        return any(_running(record) for record in records())
    def destroyed():
        return all(record['phase'] == 'destroyed' for record in records())
    def unique():
        return all(record['operation_id'] != event.get('operation_id') for record in records())
    if name == 'acquire':
        if result['lock']['state'] == 'held':
            fail('lifecycle lock held')
        result['lock'] = {'state': 'held', 'run_id': event['run_id']}
    else:
        if result['lock']['state'] != 'held':
            fail('lifecycle lock not held')
        if result['lock']['run_id'] != event['run_id']:
            fail('lifecycle owner mismatch')
        if name == 'declare':
            require(result['status'] == 'active' and not active())
            requested = _topology(event['topology'])
            require(len(set(result['nodes']) | {node['node_id'] for node in requested}) <= 1000)
            for node in result['nodes'].values():
                node['desired'] = False
            keys = state_keys(identity['profile'], [node['node_id'] for node in requested])
            for node in requested:
                old = result['nodes'].get(node['node_id'])
                result['nodes'][node['node_id']] = {**old, 'desired': True} if old and old['phase'] != 'destroyed' else {
                    'state_key': keys['nodes'][node['node_id']], 'role': node['role'], 'index': node['index'], 'desired': True, **_idle_shared()}
            result['topology_declared'] = True
        elif name == 'key-intent':
            require(result['status'] == 'active' and result['key']['phase'] == 'absent')
            result['key'] = {'mode': event['mode'], 'phase': 'intent', 'fingerprint': None}
        elif name == 'key-prepared':
            require(result['key']['phase'] == 'intent' and (_fingerprint(event['fingerprint']) if result['key']['mode'] == 'managed' else event['fingerprint'] is None))
            result['key'].update(phase='prepared', fingerprint=event['fingerprint'])
        elif name == 'shared-start':
            require(result['status'] == 'active' and result['topology_declared'] and result['key']['phase'] == 'prepared' and not active() and result['shared']['phase'] in ('declared', 'ready') and unique())
            result['shared'] = {'phase': 'running', 'operation': 'create', 'operation_id': event['operation_id']}
        elif name == 'shared-destroy':
            require(all(node['phase'] == 'destroyed' for node in result['nodes'].values()) and result['shared']['phase'] not in ('running', 'destroying', 'destroyed') and unique())
            result['shared'] = {'phase': 'destroying', 'operation': 'destroy', 'operation_id': event['operation_id']}
        elif name in ('shared-complete', 'shared-fail'):
            require(_running(result['shared']) and result['shared']['operation_id'] == event['operation_id'])
            result['shared']['phase'] = 'failed' if name == 'shared-fail' else 'ready' if result['shared']['operation'] == 'create' else 'destroyed'
        elif name == 'shared-retry':
            require(result['shared']['phase'] == 'failed' and result['shared']['operation'] == 'create')
            result['shared'] = _idle_shared()
        elif name in ('start', 'destroy'):
            require(event['node_id'] in result['nodes'] and unique())
            node = result['nodes'][event['node_id']]
            if name == 'start':
                require(result['status'] == 'active' and node['desired'] and result['key']['phase'] == 'prepared' and result['shared']['phase'] == 'ready' and node['phase'] in ('declared', 'ready'))
            else:
                require(not node['desired'] and node['phase'] not in ('running', 'destroying', 'destroyed'))
            node.update(phase='running' if name == 'start' else 'destroying', operation='create' if name == 'start' else 'destroy', operation_id=event['operation_id'])
        elif name in ('complete', 'fail'):
            require(event['node_id'] in result['nodes'])
            node = result['nodes'][event['node_id']]
            require(_running(node) and node['operation_id'] == event['operation_id'])
            node['phase'] = 'failed' if name == 'fail' else 'ready' if node['operation'] == 'create' else 'destroyed'
        elif name == 'retry':
            require(event['node_id'] in result['nodes'])
            node = result['nodes'][event['node_id']]
            require(node['phase'] == 'failed' and node['operation'] == 'create')
            node.update(_idle_shared())
        elif name == 'begin-delete':
            require(result['status'] != 'retired' and not active())
            result['status'] = 'deleting'
            for node in result['nodes'].values():
                node['desired'] = False
        elif name == 'key-cleanup':
            require(result['status'] == 'deleting' and destroyed() and result['key']['phase'] == 'prepared')
            result['key']['phase'] = 'cleanup'
        elif name == 'key-removed':
            require(result['key']['phase'] == 'cleanup')
            result['key']['phase'] = 'removed'
        elif name == 'retire':
            require(result['status'] == 'deleting' and result['key']['phase'] == 'removed' and destroyed())
            result['status'] = 'retired'
        elif name == 'recreate':
            require(result['status'] == 'retired' and not active() and result['generation'] < MAX_INTEGER)
            result.update(generation=int(result['generation']) + 1, status='active', key={'mode': None, 'phase': 'absent', 'fingerprint': None}, shared=_idle_shared(), topology_declared=False)
        elif name == 'release':
            require(not active() and result['key']['phase'] not in ('intent', 'cleanup'))
            result['lock'] = {'state': 'idle', 'run_id': None}
        else:
            require(False)
    result['revision'] = int(result['revision']) + 1
    result['write_id'] = event['write_id']
    return {'condition': {'if_match': observation['etag']}, 'document': result}

File: red/src/lifecycle.ts
import {identityValid,identityEqual} from './coordination.ts';
import {expand,state_keys} from './index.ts';
type Map=Record<string,any>;
const object=(v:unknown):v is Map=>v!==null&&typeof v==='object'&&!Array.isArray(v);
const fields=(v:unknown,keys:string[],optional:string[]=[]):v is Map=>object(v)&&keys.every(k=>Object.hasOwn(v,k))&&Object.keys(v).every(k=>[...keys,...optional].includes(k));
const safe=(v:unknown)=>typeof v==='string'&&/^[a-zA-Z0-9][a-zA-Z0-9_-]{0,62}$/.exec(v)?.[0]===v;
const nonblank=(v:unknown)=>typeof v==='string'&&!!v.trim();
const integer=(v:unknown,min:number,max=Number.MAX_SAFE_INTEGER)=>typeof v==='number'&&Number.isSafeInteger(v)&&v>=min&&v<=max;
const roleValid=(v:unknown)=>v===null||(typeof v==='string'&&/^[a-z][a-z0-9]*(-[a-z0-9]+)*$/.exec(v)?.[0]===v);
const fingerprint=(v:unknown)=>typeof v==='string'&&/^SHA256:[A-Za-z0-9+/]{43}$/.exec(v)?.[0]===v;
const running=(record:Map)=>record.phase==='running'||record.phase==='destroying';
const idleShared=()=>({phase:'declared',operation:null,operation_id:null});
function topology(value:unknown):Map[]|null {
  if(!Array.isArray(value)||!value.length||value.length>1000)return null;
  let total=0;
  for(const entry of value) {
    if(!fields(entry,[],['role','count'])||!roleValid(entry.role??null)||!integer(entry.count===undefined?1:entry.count,1,1000))return null;
    total+=entry.count??1;if(total>1000)return null;
  }
  try{return expand(value);}catch{return null;}
}
function recordValid(record:Map):boolean {
  if(!['declared','running','ready','failed','destroying','destroyed'].includes(record.phase))return false;
  if(record.phase==='declared')return record.operation===null&&record.operation_id===null;
  if(!safe(record.operation_id)||!['create','destroy'].includes(record.operation))return false;
  if(['running','ready'].includes(record.phase))return record.operation==='create';
  if(['destroying','destroyed'].includes(record.phase))return record.operation==='destroy';
  return true;
}
export function lifecycleDocumentValid(value:unknown):value is Map {
  if(!fields(value,['schema_version','identity','revision','write_id','lock','generation','status','topology_declared','key','shared','nodes'])||value.schema_version!==2||!identityValid(value.identity)||!integer(value.revision,1)||!integer(value.generation,1)||!safe(value.write_id)||!['active','deleting','retired'].includes(value.status)||typeof value.topology_declared!=='boolean')return false;
  if(!fields(value.lock,['state','run_id'])||!((value.lock.state==='held'&&safe(value.lock.run_id))||(value.lock.state==='idle'&&value.lock.run_id===null)))return false;
  const key=value.key;
  if(!fields(key,['mode','phase','fingerprint'])||!['absent','intent','prepared','cleanup','removed'].includes(key.phase))return false;
  if(key.phase==='absent') {if(key.mode!==null||key.fingerprint!==null)return false;}
  else {
    if(!['managed','external'].includes(key.mode))return false;
    if(key.mode==='external'||key.phase==='intent'){if(key.fingerprint!==null)return false;}
    else if(!fingerprint(key.fingerprint))return false;
  }
  if(!fields(value.shared,['phase','operation','operation_id'])||!recordValid(value.shared)||!object(value.nodes)||Object.keys(value.nodes).length>1000)return false;
  const ops=new Set<string>();if(value.shared.operation_id!==null)ops.add(value.shared.operation_id);
  const desiredRoles=new Map<string|null,number[]>();
  if(running(value.shared)&&value.lock.state!=='held')return false;
  for(const [id,node] of Object.entries(value.nodes)) {
    if(!safe(id)||!fields(node,['state_key','role','index','desired','phase','operation','operation_id'])||!roleValid(node.role)||!integer(node.index,0,999)||typeof node.desired!=='boolean'||!recordValid(node))return false;
    if(id!==(node.role===null?String(node.index):`${node.role}-${node.index}`)||node.state_key!==`${value.identity.profile}/compute/nodes/${id}.tfstate`)return false;
    if(node.operation_id!==null){if(ops.has(node.operation_id))return false;ops.add(node.operation_id);}
    if(running(node)&&value.lock.state!=='held')return false;
    if(node.desired)desiredRoles.set(node.role,[...(desiredRoles.get(node.role)??[]),node.index]);
  }
  const allDestroyed=[value.shared,...Object.values(value.nodes)].every((record:any)=>record.phase==='destroyed');
  const allUndesired=Object.values(value.nodes).every((node:any)=>!node.desired);
  if(value.status==='deleting'&&!allUndesired)return false;
  if(value.status==='retired'&&(value.key.phase!=='removed'||!allDestroyed||!allUndesired))return false;
  if(['cleanup','removed'].includes(value.key.phase)&&(!['deleting','retired'].includes(value.status)||!allDestroyed))return false;
  if(desiredRoles.has(null)&&desiredRoles.size!==1)return false;
  if(![...desiredRoles.values()].every(indices=>indices.sort((a,b)=>a-b).every((index,position)=>index===position)))return false;
  return true;
}
function eventValid(event:unknown):event is Map {
  if(!object(event)||typeof event.type!=='string'||!event.type.startsWith('lifecycle/'))return false;
  const name=event.type.slice(10);
  const extras:Record<string,string[]>={acquire:[],release:[],'begin-delete':[],retire:[],recreate:[],declare:['topology'],'key-intent':['mode'],'key-prepared':['fingerprint'],'key-cleanup':[],'key-removed':[],'shared-start':['operation_id'],'shared-destroy':['operation_id'],'shared-complete':['operation_id'],'shared-fail':['operation_id'],'shared-retry':['evidence'],start:['node_id','operation_id'],destroy:['node_id','operation_id'],complete:['node_id','operation_id'],fail:['node_id','operation_id'],retry:['node_id','evidence']};
  if(!Object.hasOwn(extras,name)||!fields(event,['type','run_id','write_id','target_etag',...extras[name]])||!safe(event.run_id)||!safe(event.write_id)||!(event.target_etag===null||nonblank(event.target_etag)))return false;
  if(Object.hasOwn(event,'node_id')&&!safe(event.node_id))return false;
  if(Object.hasOwn(event,'operation_id')&&!safe(event.operation_id))return false;
  if(name==='declare'&&!topology(event.topology))return false;
  if(name==='key-intent'&&!['managed','external'].includes(event.mode))return false;
  if(name==='key-prepared'&&event.fingerprint!==null&&!fingerprint(event.fingerprint))return false;
  if(Object.hasOwn(event,'evidence')&&event.evidence!=='readable-state'&&!(['shared-retry','retry'].includes(name)&&event.evidence==='verified-provider-absence'))return false;
  return true;
}
function fail(message:string):never{throw new Error(message);}
function requireTransition(condition:unknown):asserts condition{if(!condition)fail('lifecycle transition refused');}

/** Schema-2 transition plan only; transport must confirm each intent. */
export function lifecycle(observation:unknown,identity:unknown,event:unknown) {
  if(!identityValid(identity))fail('invalid lifecycle identity');
  if(!eventValid(event))fail('invalid lifecycle event');
  if(!object(observation)||!(['absent','error'].includes(observation.status)?fields(observation,['status']):observation.status==='present'&&fields(observation,['status','etag','document'])&&nonblank(observation.etag)))fail('invalid lifecycle observation');
  if(observation.status==='error')fail('lifecycle read failed');
  const present=observation.status==='present',doc=observation.document,name=event.type.slice(10);
  if(present&&!lifecycleDocumentValid(doc))fail('invalid lifecycle document');
  if(present&&!identityEqual(doc.identity,identity))fail('lifecycle identity mismatch');
  if(event.target_etag!==(present?observation.etag:null))fail('lifecycle stale observation');
  if(present&&event.write_id===doc.write_id)fail('lifecycle write_id reused');
  if(present&&doc.revision===Number.MAX_SAFE_INTEGER)fail('lifecycle revision exhausted');
  if(!present) {
    requireTransition(name==='acquire');
    return {condition:{if_none_match:'*'},document:{schema_version:2,identity:structuredClone(identity),revision:1,write_id:event.write_id,lock:{state:'held',run_id:event.run_id},generation:1,status:'active',topology_declared:false,key:{mode:null,phase:'absent',fingerprint:null},shared:idleShared(),nodes:{}}};
  }
  const next:Map=structuredClone(doc);const records=()=>[next.shared,...Object.values(next.nodes)] as Map[];
  const activeWork=()=>records().some(running);const allDestroyed=()=>records().every(record=>record.phase==='destroyed');
  const unique=()=>records().every(record=>record.operation_id!==event.operation_id);
  if(name==='acquire') {
    if(next.lock.state==='held')fail('lifecycle lock held');next.lock={state:'held',run_id:event.run_id};
  }else {
    if(next.lock.state!=='held')fail('lifecycle lock not held');if(next.lock.run_id!==event.run_id)fail('lifecycle owner mismatch');
    switch(name) {
      case 'declare': {
        requireTransition(next.status==='active'&&!activeWork());
        const requested=topology(event.topology)!;const ids=new Set(requested.map(node=>node.node_id));
        requireTransition(new Set([...Object.keys(next.nodes),...ids]).size<=1000);
        for(const node of Object.values(next.nodes) as Map[])node.desired=false;
        const keys=state_keys(identity.profile,requested.map(node=>node.node_id));
        for(const node of requested){
          const old=next.nodes[node.node_id];
          next.nodes[node.node_id]=old&&old.phase!=='destroyed'?{...old,desired:true}:{state_key:keys.nodes[node.node_id],role:node.role,index:node.index,desired:true,...idleShared()};
        }
        next.topology_declared=true;break;
      }
      case 'key-intent':requireTransition(next.status==='active'&&next.key.phase==='absent');next.key={mode:event.mode,phase:'intent',fingerprint:null};break;
      case 'key-prepared':requireTransition(next.key.phase==='intent'&&(next.key.mode==='managed'?fingerprint(event.fingerprint):event.fingerprint===null));next.key.phase='prepared';next.key.fingerprint=event.fingerprint;break;
      case 'shared-start':requireTransition(next.status==='active'&&next.topology_declared&&next.key.phase==='prepared'&&!activeWork()&&['declared','ready'].includes(next.shared.phase)&&unique());next.shared={phase:'running',operation:'create',operation_id:event.operation_id};break;
      case 'shared-destroy':requireTransition((Object.values(next.nodes) as Map[]).every(node=>node.phase==='destroyed')&&!['running','destroying','destroyed'].includes(next.shared.phase)&&unique());next.shared={phase:'destroying',operation:'destroy',operation_id:event.operation_id};break;
      case 'shared-complete':case 'shared-fail':requireTransition(running(next.shared)&&next.shared.operation_id===event.operation_id);next.shared.phase=name==='shared-fail'?'failed':next.shared.operation==='create'?'ready':'destroyed';break;
      case 'shared-retry':requireTransition(next.shared.phase==='failed'&&next.shared.operation==='create');next.shared=idleShared();break;
      case 'start':case 'destroy': {
        requireTransition(Object.hasOwn(next.nodes,event.node_id)&&unique());const node=next.nodes[event.node_id];
        if(name==='start')requireTransition(next.status==='active'&&node.desired&&next.key.phase==='prepared'&&next.shared.phase==='ready'&&['declared','ready'].includes(node.phase));
        else requireTransition(!node.desired&&!['running','destroying','destroyed'].includes(node.phase));
        node.phase=name==='start'?'running':'destroying';node.operation=name==='start'?'create':'destroy';node.operation_id=event.operation_id;break;
      }
      case 'complete':case 'fail': {requireTransition(Object.hasOwn(next.nodes,event.node_id));const node=next.nodes[event.node_id];requireTransition(running(node)&&node.operation_id===event.operation_id);node.phase=name==='fail'?'failed':node.operation==='create'?'ready':'destroyed';break;}
      case 'retry': {requireTransition(Object.hasOwn(next.nodes,event.node_id));const node=next.nodes[event.node_id];requireTransition(node.phase==='failed'&&node.operation==='create');Object.assign(node,idleShared());break;}
      case 'begin-delete':requireTransition(next.status!=='retired'&&!activeWork());next.status='deleting';for(const node of Object.values(next.nodes) as Map[])node.desired=false;break;
      case 'key-cleanup':requireTransition(next.status==='deleting'&&allDestroyed()&&next.key.phase==='prepared');next.key.phase='cleanup';break;
      case 'key-removed':requireTransition(next.key.phase==='cleanup');next.key.phase='removed';break;
      case 'retire':requireTransition(next.status==='deleting'&&next.key.phase==='removed'&&allDestroyed());next.status='retired';break;
      case 'recreate':requireTransition(next.status==='retired'&&!activeWork()&&next.generation<Number.MAX_SAFE_INTEGER);next.generation++;next.status='active';next.key={mode:null,phase:'absent',fingerprint:null};next.shared=idleShared();next.topology_declared=false;break;
      case 'release':requireTransition(!activeWork()&&!['intent','cleanup'].includes(next.key.phase));next.lock={state:'idle',run_id:null};break;
      default:requireTransition(false);
    }
  }
  next.revision++;next.write_id=event.write_id;
  return {condition:{if_match:observation.etag},document:next};
}
